File: Language/C#/Book/tmp/LINQ实战/LinqInAction/Chapter12.CSharp/LinqToAmazon/AmazonBookSearch.cs
#region Imports

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;

#endregion Imports
  
namespace LinqInAction.Chapter12.LinqToAmazon
{
  public class AmazonBookSearch : IEnumerable<AmazonBook>
  {
    private AmazonBookQueryCriteria _Criteria;

    #region Query operators

    public AmazonBookSearch Where(
      Expression<Func<AmazonBook, Boolean>> predicate)
    {
      _Criteria = new AmazonBookExpressionVisitor().ProcessExpression(predicate);
      return this;
    }

    public AmazonBookSearch Select<TResult>(
      Expression<Func<AmazonBook, TResult>> selector)
    {
      return this;
    }

    #endregion Query operators

    #region IEnumerable<T> Members

    IEnumerator<AmazonBook> IEnumerable<AmazonBook>.GetEnumerator()
    {
      return (IEnumerator<AmazonBook>)((IEnumerable)this).GetEnumerator();
    }

    #endregion IEnumerable Members

    #region IEnumerable Members

    IEnumerator IEnumerable.GetEnumerator()
    {
      // Execute query
      String url = AmazonHelper.BuildUrl(_Criteria);
      IEnumerable<AmazonBook> books = AmazonHelper.PerformWebQuery(url);

      return books.GetEnumerator();
    }

    #endregion IEnumerable Members
  }
}

File: Language/C#/Book/LINQ实战/第12章 扩展LINQ/Chapter12.VB/LinqToAmazon/AmazonBookSearch.vb
Imports System.Linq.Expressions

Public Class AmazonBookSearch
  Implements IEnumerable(Of AmazonBook)
  Private _Criteria As AmazonBookQueryCriteria

#Region "Query operators"

  Public Function Where(ByVal predicate As Expression(Of Func(Of AmazonBook, Boolean))) As AmazonBookSearch
    _Criteria = New AmazonBookExpressionVisitor().ProcessExpression(predicate)
    Return Me
  End Function

  Public Function [Select](Of TResult)(ByVal selector As Expression(Of Func(Of AmazonBook, TResult))) As AmazonBookSearch
    Return Me
  End Function

#End Region

#Region "IEnumerable Members"

  Private Function GetEnumerator() As IEnumerator(Of AmazonBook) Implements IEnumerable(Of AmazonBook).GetEnumerator
    Return DirectCast(DirectCast(Me, IEnumerable).GetEnumerator(), IEnumerator(Of AmazonBook))
  End Function

#End Region

#Region "IEnumerable Members"

  Private Function GetEnumeratorNonGeneric() As IEnumerator Implements IEnumerable.GetEnumerator
    ' Execute query
    Dim url As String = AmazonHelper.BuildUrl(_Criteria)
    Dim books As IEnumerable(Of AmazonBook) = AmazonHelper.PerformWebQuery(url)

    Return books.GetEnumerator()
  End Function

#End Region
End Class